modules/peers.js

```diff
@@ -421,73 +447,72 @@ Peers.prototype.acceptable = function (peers) {
 
 /**
  * Gets peers list and calculated consensus.
- * @param {Object} options - Constains limit, broadhash.
+ * @param {Object} options
+ * @param {number} options.limit[=constants.maxPeers] - maximum number of peers to get
+ * @param {string} options.broadhash[=null] - broadhash to match peers by
+ * @param {Array} options.allowedStates[=[2]] - allowed peer states
+ * @param {number} options.attempt[=undefined] - if 0: return peers with equal options.broadhash
+ *                                               if 1: return peers with different options.broadhash
+ *                                               if not specified: return peers regardless of options.broadhash
  * @param {function} cb - Callback function.
  * @returns {setImmediateCallback} error | peers, consensus
  */
 Peers.prototype.list = function (options, cb) {
-	options.limit = options.limit || constants.maxPeers;
-	options.broadhash = options.broadhash || modules.system.getBroadhash();
-	options.allowedStates = options.allowedStates || [Peer.STATE.CONNECTED];
-	options.attempts = ['matched broadhash', 'unmatched broadhash'];
-	options.attempt = 0;
-	options.matched = 0;
-
-	function randomList (options, peers, cb) {
+
+	var limit = options.limit || constants.maxPeers;
+	var broadhash = options.broadhash || modules.system.getBroadhash();
+	var allowedStates = options.allowedStates || [Peer.STATE.CONNECTED];
+	var attempts = (options.attempt === 0 || options.attempt === 1) ? [options.attempt] : [1, 0];
+
+	var attemptsDescriptions = ['matched broadhash', 'unmatched broadhash'];
+
+	function randomList (peers, cb) {
 		// Get full peers list (random)
-		__private.getByFilter ({}, function (err, peersList) {
+		__private.getByFilter({}, function (err, peersList) {
 			var accepted, found, matched, picked;
 
 			found = peersList.length;
+			var attempt = attempts.pop();
 			// Apply filters
 			peersList = peersList.filter(function (peer) {
-				if (options.broadhash) {
+				if (broadhash) {
 					// Skip banned and disconnected peers (state 0 and 1)
-					return options.allowedStates.indexOf(peer.state) !== -1 && (
+					return allowedStates.indexOf(peer.state) !== -1 && (
 						// Matched broadhash when attempt 0
-						options.attempt === 0 ? (peer.broadhash === options.broadhash) :
+						attempt === 0 ? (peer.broadhash === broadhash) :
 						// Unmatched broadhash when attempt 1
-						options.attempt === 1 ? (peer.broadhash !== options.broadhash) : false
+						attempt === 1 ? (peer.broadhash !== broadhash) : false
 					);
 				} else {
 					// Skip banned and disconnected peers (state 0 and 1)
-					return options.allowedStates.indexOf(peer.state) !== -1;
+					return allowedStates.indexOf(peer.state) !== -1;
 				}
 			});
 			matched = peersList.length;
 			// Apply limit
-			peersList = peersList.slice(0, options.limit);
+			peersList = peersList.slice(0, limit);
 			picked = peersList.length;
 			accepted = self.acceptable(peers.concat(peersList));
```

Comment:
faster - `peers.concat(self.acceptable(peersList));`, because `peers` already passed through `self.acceptable`

Reply:
Part of the `peers.list` refactor.